cloud/gke.go

```diff
@@ -0,0 +1,1038 @@
+package cloud
+
+import (
+	"os"
+	"io/ioutil"
+	"golang.org/x/oauth2/google"
+	gke "google.golang.org/api/container/v1"
+	"golang.org/x/net/context"
+	"strings"
+	"time"
+	"google.golang.org/api/googleapi"
+	"net/http"
+	"github.com/gin-gonic/gin"
+
+	banzaiUtils "github.com/banzaicloud/banzai-types/utils"
+	banzaiConstants "github.com/banzaicloud/banzai-types/constants"
+	banzaiTypes "github.com/banzaicloud/banzai-types/components"
+	banzaiSimpleTypes "github.com/banzaicloud/banzai-types/components/database"
+	"github.com/banzaicloud/banzai-types/database"
+	"encoding/base64"
+	"fmt"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/api/core/v1"
+	"k8s.io/api/rbac/v1beta1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"gopkg.in/yaml.v2"
+	"path/filepath"
+	"github.com/Azure/go-autorest/autorest/to"
+)
+
+var credentialPath string
+
+const (
+	statusRunning = "RUNNING"
+)
+
+const (
+	defaultNamespace = "default"
+	clusterAdmin     = "cluster-admin"
+	netesDefault     = "netes-default"
+)
+
+const googleAppCredential = "GOOGLE_APPLICATION_CREDENTIALS"
+
+func init() {
+	credentialPath = os.Getenv(googleAppCredential)
+	banzaiUtils.LogDebugf(banzaiConstants.TagInit, "%s is %s", googleAppCredential, credentialPath)
+}
+
+func CreateClusterGoogle(request *banzaiTypes.CreateClusterRequest, c *gin.Context) (bool, *banzaiSimpleTypes.ClusterSimple) {
+	data, err := ioutil.ReadFile(credentialPath)
+	if err != nil {
+		banzaiUtils.LogFatalf(banzaiConstants.TagCreateCluster, "GOOGLE_APPLICATION_CREDENTIALS env var is not specified: %s", err)
+		SetResponseBodyJson(c, http.StatusInternalServerError, gin.H{
+			JsonKeyStatus:  http.StatusInternalServerError,
+			JsonKeyMessage: "GOOGLE_APPLICATION_CREDENTIALS env var is not specified",
+		})
+		return false, nil
+	}
+
+	svc, err := GetGoogleServiceClient()
+	if err != nil {
+		SetResponseBodyJson(c, http.StatusInternalServerError, gin.H{
+			JsonKeyStatus:  http.StatusInternalServerError,
+			JsonKeyMessage: err,
+		})
+		return false, nil
+	}
+
+	cc := GKECluster{
+		ProjectID:         request.Properties.CreateClusterGoogle.Project,
+		Zone:              request.Location,
+		Name:              request.Name,
+		NodeCount:         int64(request.Properties.CreateClusterGoogle.Node.Count),
+		CredentialPath:    credentialPath,
+		CredentialContent: string(data),
+		MasterVersion:     request.Properties.CreateClusterGoogle.Master.Version,
+		NodeVersion:       request.Properties.CreateClusterGoogle.Node.Version,
+	}
+
+	banzaiUtils.LogInfof(banzaiConstants.TagCreateCluster, "Cluster request: %v", generateClusterCreateRequest(cc))
```

Comment:
Maybe this is nothing but why are we calling generateClusterCreateRequest(cc) three times here?